### ISSUE TITLE
NuGet.Packaging assemblies don't load for Feed package publishing

### ISSUE THREAD

**Issue by wli3** [dotnet/arcade#1965]
The latest nuget package has a breaking change removing NuGet.Packaging.Core, and moving types to NuGet.Packaging. sdk should be the first one taking that change. As a result, arcade failed to publish.

failed publish
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=2379512

**Comment by wli3** [comment]
Maybe the title should be update to latest nuget version.

**Comment by jcagme** [comment]
Hey @wli3, help me understand your process.

* Are you restoring the Arcade SDK, just the Arcade version of the `Microsoft.DotNet.Build.Tasks.Feed` package or is this the BuildTools version of `Microsoft.DotNet.Build.Tasks.Feed`?

Depending on which of this you are using we'll know what's the best followy up is.

**Comment by wli3** [comment]
only arcade version

**Comment by jcagme** [comment]
What version of the package are you using?

**Comment by wli3** [comment]
Change the title to use nuget 5.0.0-rtm.5821 and up. That is the root cause of the problem. NuGet has a breaking change to moved the package _nuget.packaging.core_

**Comment by MattGal** [comment]
@johntortugo can you explain why you put this in the FR epic?

**Comment by JohnTortugo** [comment]
Not sure @MattGal . Today it looks more like an Arcade issue though.

**Comment by MattGal** [comment]
OK with no FR details or followup for us, I am removing the epic tag;  please ping me if you feel this was in error.

**Comment by markwilkie** [comment]
@jcagme  - what are the next steps here?

**Comment by jcagme** [comment]
I don't know what's the latest from @wli3. We could update the version although I don't know why would we take a dependency on a RTM package.

**Comment by markwilkie** [comment]
@wli3 - thoughts on next steps?

**Comment by markwilkie** [comment]
Closing based on lack of responses

**Comment by wli3** [comment]
Sorry for the delay. CLI and SDK repo workaround this issue with painful product change. However, nuget version past 5.0.0-rtm.5821 has a breaking change of moving types. It might still cause problem like build task failure or cross gen that is similar to CLI and SDK repo. So I suggest Arcade to upgrade ASAP in case the problem appear before final release

**Comment by wtgodbe** [comment]
Reopening as we just got bitten by this in AspNetCore-Tooling: https://github.com/dotnet/aspnetcore/issues/18324.

@wli3 can you comment on how you worked around this in CLI/SDK?

@markwilkie @jcagme can we update to a newer version of nuget.client to unblock upstack repos? I think anyone using a new SDK with a new Arcade will be broken. Here's the call that fails: https://github.com/dotnet/aspnetcore/issues/18324

**Comment by nkolev92** [comment]
@wtgodbe Is arcade depending on another component that depends on NuGet, or it depends on NuGet directly. 

If it's the first one, you can add NuGet.Packaging.Core as a reference, it has type forwarding. 
If it's the latter one, then simply rebuilding against a newer version of NuGet should do it.

**Comment by markwilkie** [comment]
Assuming Arcade, the .net sdk needs to be updated to get a new nuget.exe.  I seem to recall that this happened already (but I don't know for sure).  So perhaps this build is getting nuget from elsewhere somehow?  Or, perhaps the sdk doesn't have the latest one?

cc/ @mmitche

**Comment by wtgodbe** [comment]
`Toolset` & `SDK` use a 5.5.0-preview.2 version of NuGet.Build.Tasks, from 3 weeks ago:
https://github.com/dotnet/toolset/blob/master/eng/Version.Details.xml#L72-L74
https://github.com/dotnet/sdk/blob/master/eng/Version.Details.xml#L22-L24

Whereas Arcade uses Nuget.Versioning version 4.4.0, Nuget.Client version 5.3.0: https://github.com/dotnet/arcade/blob/be2d4ae34b4fcfe35beb9bdaeb7b7990c1147740/eng/Versions.props#L42-L43. So I believe we'd need to update Arcade in this repo as @nkolev92 suggested

**Comment by wtgodbe** [comment]
Nuget.Versioning 4.4.0 is from 2017 - it appears we use that version because anything newer would conflict with the MSBuild version used by the SDK: https://github.com/dotnet/arcade/pull/2189. The SDK now uses MSBuild version `16.6.0-preview-20064-05`:https://github.com/dotnet/toolset/blob/30eddcbf073098f8958a0a0321aad4a6e3ad3b79/eng/Version.Details.xml#L40. MSbuild uses a `5.4.0` version of Nuget: https://github.com/microsoft/msbuild/blob/ab9b2f36a5ff7a85f842b205d5529e77fdc9d7ab/eng/Packages.props#L4. So I believe we can bump the Nuget.Versioning version up to at least 5.3.0, and see if that resolves the issue. I'll try this all out with an internal branch.

**Comment by wtgodbe** [comment]
Here's the internal Arcade build with my branch: https://dev.azure.com/dnceng/internal/_build/results?buildId=484408

**Comment by wli3** [comment]
Last time nuget added type forwarding, so it passed. I don't know the cause of this time.

**Comment by markwilkie** [comment]
I see - yes, we can (and should) update these versions.  @mmitche, any thoughts on not having more than one place where we're defining the nuget version?

**Comment by mmitche** [comment]
> I see - yes, we can (and should) update these versions. @mmitche, any thoughts on not having more than one place where we're defining the nuget version?

I'm not sure how we would get the SDK's nuget version...but primarily I was concerned about two copies of nuget.exe itself. Maybe @tmat has some ideas here?

**Comment by jaredpar** [comment]
@wtgodbe where is a link to the actual fix that you built that off of? Looking through the thread I just see a build, not a PR.

**Comment by wtgodbe** [comment]
[Here](https://dev.azure.com/dnceng/internal/_git/d551d0d5-053b-400b-99d7-56e9c6c469c4/commit/1c552e8f00a4cf5ed104e6b8c5355c58d0542564) it is, I just bumped `NuGetVersioningVersion` to `5.3.0`

**Comment by wtgodbe** [comment]
Updating the `NugetVersioningVersion` in Arcade didn't work in my internal build: d2bd58c62f9104d66f415141a1a27b665c78690c. Still getting the same error:

> D:\a\1\s.packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20065.1\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error : Method not found: 'NuGet.Versioning.NuGetVersion NuGet.Packaging.Core.PackageIdentity.get_Version()'.

I'll try adding a direct reference to Nuget.Packaging.Core

**Comment by wtgodbe** [comment]
Trying again, this time adding a direct reference to `Nuget.Package.Core` in `Microsoft.DotNet.Build.Tasks.Feed`: https://dev.azure.com/dnceng/internal/_build/results?buildId=484708

**Comment by wtgodbe** [comment]
And the new AspNetCore-Tooling build: https://dev.azure.com/dnceng/internal/_build/results?buildId=484794

**Comment by wtgodbe** [comment]
Nope, that failed too. @nkolev92 is there something else I should try? I tried updating this repo's `Nuget.Versioning` version to 5.3.0, and adding a packageReference to `Nuget.Packaging.Core` in the failed project, but neither worked (See above comments for links to the builds)

**Comment by nkolev92** [comment]
The project in question, already had nuget.packaging.core and nuget.packaging as dependencies and the correct version of those. 

Specifically: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Build.Tasks.Feed/Microsoft.DotNet.Build.Tasks.Feed.csproj has a project reference to https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.VersionTools/lib/Microsoft.DotNet.VersionTools.csproj which depends on nuget.packaging and nuget.packaging.

Do we know where this call is actually coming from?

**Comment by wtgodbe** [comment]
The call happens here: https://github.com/dotnet/arcade/blob/be2d4ae34b4fcfe35beb9bdaeb7b7990c1147740/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs#L994-L996

We invoke the task directly from https://github.com/dotnet/arcade/blob/master/eng/common/sdk-task.ps1

With the following command (taken from one of the failing builds):

> Formatted command: . 'd:\a\1\s\eng\common\sdk-task.ps1' -task PublishArtifactsInManifest -restore -msbuildEngine dotnet /p:ArtifactsCategory=.NETCORE /p:IsStableBuild=False /p:IsInternalBuild=False /p:RepositoryName=dotnet-aspnetcore-tooling /p:CommitSha=8f34f18e0d5bfc8701d93a106acf707fb44f9ca8 /p:NugetPath=C:\hostedtoolcache\windows\NuGet\5.4.0\x64\nuget.exe /p:AzdoTargetFeedPAT='***' /p:AzureStorageTargetFeedPAT='***' /p:BARBuildId=39995 /p:MaestroApiEndpoint='https://maestro-prod.westus2.cloudapp.azure.com' /p:BuildAssetRegistryToken='***' /p:ManifestsBasePath='d:\a\1\a/AssetManifests/' /p:BlobBasePath='d:\a\1\a/BlobArtifacts/' /p:PackageBasePath='d:\a\1\a/PackageArtifacts/' /p:Configuration=Release /p:PublishInstallersAndChecksums=true /p:InstallersTargetStaticFeed=https://dotnetcli.blob.core.windows.net/dotnet/index.json /p:InstallersAzureAccountKey=*** /p:ChecksumsTargetStaticFeed=https://dotnetclichecksums.blob.core.windows.net/dotnet/index.json /p:ChecksumsAzureAccountKey=*** /p:AzureDevOpsStaticShippingFeed='https://pkgs.dev.azure.com/dnceng/public/_packaging/general-testing/nuget/v3/index.json' /p:AzureDevOpsStaticShippingFeedKey='***' /p:AzureDevOpsStaticTransportFeed='https://pkgs.dev.azure.com/dnceng/public/_packaging/general-testing/nuget/v3/index.json' /p:AzureDevOpsStaticTransportFeedKey='***' /p:AzureDevOpsStaticSymbolsFeed='https://pkgs.dev.azure.com/dnceng/public/_packaging/general-testing-symbols/nuget/v3/index.json' /p:AzureDevOpsStaticSymbolsFeedKey='***'

**Comment by markwilkie** [comment]
This *should* be using the nuget client from the sdk itself.  @JohnTortugo - any insight into what could be happening here?

**Comment by wtgodbe** [comment]
ping @JohnTortugo - AspNetCore-Tooling is currently in a very broken state because of this issue

**Comment by markwilkie** [comment]
We're looking at this even more now @wtgodbe .  Which releases is this currently blocking/affecting so we can set the right level of urgency?

**Comment by wtgodbe** [comment]
It's only blocking in 5.0, but it's the last thing stopping us from ingesting dotnet/runtime builds, which are currently several months out of date for us

**Comment by markwilkie** [comment]
Thanks @wtgodbe.  yea, a big deal for sure.

**Comment by JohnTortugo** [comment]
I'm looking into it right now.

**Comment by JohnTortugo** [comment]
Moving this to FR so that someone can dig deeper into this. 

Here are my notes/findings & possible next steps:

- @nkolev92 suggested directly referencing Nuget.Packaging.Core as a solution
- @wtgodbe created a new Tasks.Feed that directly reference Nuget.Packaging.Core 5.3.0 but that didn't solve the issue.

- I tried reproducing the failure locally (using latest and also 5.0.100-alpha.1.20063.3 sdk)  but couldn't. I tried using an Console App. Perhaps a better approach is trying using a MSBuild Target/Task.

**Comment by wtgodbe** [comment]
@JohnTortugo @riarenas @markwilkie has anyone had a chance to look at this today?

**Comment by JohnTortugo** [comment]
I could repro it locally with the steps I mentioned in the previous comment: using an MSBuild Target/task. I'll expend some time looking for a fix.

**Comment by JohnTortugo** [comment]
The issue seems to be related to the use of DotNet.SleetLib; That package reference an older version of NuGet.Packaging and I think that's somehow causing the problem.

**Comment by wtgodbe** [comment]
The same thing is happening in the dotnet/sdk repo: https://dev.azure.com/dnceng/internal/_build/results?buildId=489383

**Comment by chcosta** [comment]
Looking

**Comment by wtgodbe** [comment]
Here's the last time we updated Sleet: https://github.com/dotnet/arcade/pull/4340

As a benchmark, dotnet/extensions uses an SDK from previous to this Arcade change. On nuget.org, both the new & old sleetlib versions depend on nuget.packaging >= 4.3.0

**Comment by chcosta** [comment]
Thanks @wtgodbe.  I'm setting up a repro now (playing catch up).  I'll reach out to you when I'm in a better position to investigate (hopefully this afternoon)

**Comment by markwilkie** [comment]
For all on this thread - we are treating this as the highest priority right now as we understand the impact.

**Comment by chcosta** [comment]
I have a local repro.  Sleet might be a red herring.  

It appears that the conflict is between Arcade's feed package using NuGet v 5.3.0.4 and dotnet loading v5.5.0.1.  I don't see any older versions of NuGet being loaded.  I'm continuing to investigate.

**Comment by wtgodbe** [comment]
The SDK first updated it's nuget dependency to 5.5 in december, which lines up with the timeline of when the bug started occuring (extensions uses an SDK from november and doesn't have this issue): https://github.com/dotnet/toolset/pull/3735

**Comment by chcosta** [comment]
Yes, I don't have a solution yet.  The failure does correspond with that change.  There is a conflict between the NuGet which comes with the Arcade Feed package and the one in the dotnet sdk.  Updating Arcade to the same version of NuGet doesn't fix the issue (I'd presume some types moved and an additional package reference is needed to make the Arcade Feed package work).  Removing NuGet assemblies from the Feed package seems to allow .net core to load from its sdk and things work.... I'm still investigating, and don't have an optimal solution at the moment.  @nguerrera , does this sound reasonable?  

I wonder if we can build the feed package and exclude NuGet assemblies from the package on core.

**Comment by nguerrera** [comment]
cc @dsplaisted. 

@chcosta I didn't follow what exactly that entails. Do you have a change we can review?

**Comment by nguerrera** [comment]
Are  you saying you can remove your dependency on NuGet altogether?

**Comment by chcosta** [comment]
Update, just saw this info on my GitHub page and realized I hadn't hit the "Comment" button like I intended to yesterday.  Posting this, and then I'll add an additional update with the latest.

> Our version of Sleet was compiled against NuGet API v4.4.0.  NuGet made a change where “PackageIdentity” moved from the “NuGet.Packaging.Core” assembly to “NuGet.Packaging”.  

> Next steps:
> 1.	@emgarten believes that we could, fairly “easily”, merge in sleet master to roll us forward to NuGet version 5.2 and from there make the update to version 5.4.  This should, theoretically” fix the problem because we will start looking for PackageIdentity in the correct assembly.

> 2.	worth with @jcagme to get a version of sleetlib published with the update

> 3.	@rainersigwald  believes that https://github.com/microsoft/msbuild/pull/4916, will “fix” the issue going forward such that we don’t hit this because we shouldn’t be mixing assemblies.  This will require dotnet/toolset to update to a version of MSBuild that contains the fix.

> @nguerrera , sorry for the slow response, it was a busy day.  I'll reach out to you tomorrow because not all the details are particularly clear yet.  I just send out mail with slightly more info

**Comment by chcosta** [comment]
@nguerrera noticed that the sdk repo is already building with microsoft/msbuild#4916 and that corresponds with the symptoms being seen here.  Assembly isolation is likely causing this problem since not all NuGet packages are being loaded.

The quickest fix would be to update `global.json` to use a version of the sdk that does not contain microsoft/msbuild#4916.  The most recent version I could find is `5.0.100-alpha1-016143`.  I have a mail out asking if that is a sane path forward or if there are alternative paths to unblock.  Given the symptoms, I don't think that an update to Sleet will fix the issue we are seeing.

**Comment by markwilkie** [comment]
For folks following this issue, the conversation somehow got onto email.  (hence the lack of apparent activity on this issue).  Currently @rainersigwald and @chcosta are working through both the closest path to unblock as well as the longer term solution.  

This is a very high priority item.

**Comment by chcosta** [comment]
I'm in the process of trying to validate an interim solution using an updated Arcade SDK in a dotnet-sdk build.  Unfortunately, I don't think there's an easy way to validate the publishing step itself apart from an update to master, so the current validation build is more of a sanity check.  I'll prepare a PR with the interim solution.

**Comment by chcosta** [comment]
Working with @wtgodbe to validate the change in an aspnetcore build.

**Comment by chcosta** [comment]
Workaround available: Arcade SDK `5.0.0-beta.20074.3` from  https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json

@wtgodbe ran a build of aspnetcore-tooling that used the one-off Arcade SDK I produced and it appears to be a viable workaround.

I actually think it's probably fine for us to just make this change in Arcade (even though it's a bit of a hack).  I'll submit a PR to Arcade but anyone that needs to unblock themselves can use the Arcade version listed above.

Meanwhile, the current theory is that MSBuild is not loading some assemblies because of its strict version matching.  @rainersigwald is continuing to investigate.

**Comment by wtgodbe** [comment]
Woohoo! Once the workaround is in Arcade I'll make sure the new version gets pushed to aspnetcore-tooling/sdk

**Comment by chcosta** [comment]
Sounds good, also, there's no reason you can't use the workaround version as I produced that from an official build (not my local dev box).  whichever is necessary to unblock you.

**Comment by wtgodbe** [comment]
@chcosta doesn't appear to have worked in master: https://dev.azure.com/dnceng/internal/_build/results?buildId=496713

**Comment by chcosta** [comment]
@wtgodbe, can you use the version I listed above? That's the version I validated. The other version merged into master *should * be the same, but I haven't validated it yet. 😕

**Comment by chcosta** [comment]
Additionally, it looks like my change is in the 5.0.0-beta.20074.6 arcade sdk but you got the build before. Do now that I still haven't validated the package contents yet though

**Comment by chcosta** [comment]
Looks like the build is passing with the latest arcade sdk https://dev.azure.com/dnceng/internal/_build/results?buildId=496755&view=results
